llvm/lib/Target/EVM/EVMAsmPrinter.cpp

```diff
@@ -105,6 +125,49 @@ bool EVMAsmPrinter::isBlockOnlyReachableByFallthrough(
   return false;
 }
 
+void EVMAsmPrinter::emitLinkerSymbol(const MachineInstr *MI) {
+  MCSymbol *LinkerSymbol = MI->getOperand(0).getMCSymbol();
+  StringRef LinkerSymbolName = LinkerSymbol->getName();
+  unsigned Opc = MI->getOpcode();
+
+  std::string SymbolNameHash = EVM::getLinkerSymbolHash(LinkerSymbolName);
+  std::string DataSymbolNameHash =
+      (Opc == EVM::DATASIZE_S) ? EVM::getDataSizeSymbol(SymbolNameHash)
+                               : EVM::getDataOffsetSymbol(SymbolNameHash);
+
+  MCInst MCI;
+  if (Opc == EVM::DATASIZE_S || Opc == EVM::DATAOFFSET_S)
+    MCI.setOpcode(EVM::PUSH4_S);
+
+  MCSymbolRefExpr::VariantKind Kind = MCSymbolRefExpr::VariantKind::VK_EVM_DATA;
+  MCOperand MCOp = MCOperand::createExpr(
+      MCSymbolRefExpr::create(DataSymbolNameHash, Kind, OutContext));
+  MCI.addOperand(MCOp);
+  EmitToStreamer(*OutStreamer, MCI);
+
+  // The linker symbol and the related section already exist, so just exit.
+  if (LinkerSymbolSet.contains(SymbolNameHash))
+    return;
+
+  LinkerSymbolSet.insert(SymbolNameHash);
+
+  MCSection *CurrentSection = OutStreamer->getCurrentSectionOnly();
+
+  // Emit the .linker_symbol_name section that contains the actual symbol
```

Comment:
Can we emit this section at the end of a file?
Currently this section is generated in the .text section:
```
foo:                                    ; @foo
; %bb.0:                                ; %entry
	JUMPDEST
	PUSH4 @__dataoffset___$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__@evm_data
	.linker_symbol_name__$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__
	.ascii	"D_105_deployed"
	.text
	PUSH4 @__datasize___$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__@evm_data
	ADD
	DUP2
	DUP2
	SWAP3
	POP
	SWAP1
	POP
	JUMP
```

Reply:
Sure, we can move it to the end. This just looks, as if we break `.text `section, but actually this syntax is fine. Here we just  switch from` .text ` to `.linker_symbol_name__$0336fd807c0716a535e520df5b63ecc41ba7984875fdfa2241fcf3c8d0107e26$__` section and then switch back to the ` .text `.
(Strictly speaking, the `.section` directive is missing before `.linker_symbol_name...`, because this is a custom section).
But I agree let's just remove it now and restore when we start using these sections.